### ISSUE TITLE
[windows support] setting a garbage collector currently causes compiled programs to crash on startup

### ISSUE THREAD

**Issue by Keshi** [scala-native/scala-native#893]
Garbage collection doesn't seem to be working on windows at the moment.  Empty programs built with scala native crash with "Could not create file mapping object (1455)" printed to console at some point upon start up if nativeGC:="immix" or nativeGC:="boehm" is selected.   MSDN says that 1455 is ERROR_COMMITMENT_LIMIT, which means "The paging file is too small for this operation to complete".  
However, if built with nativeGC:="none", things run fine.

**Comment by muxanick** [comment]
what windows version do you use?

**Comment by Keshi** [comment]
Oh, this is quite cool!  I'm able to run this in the visual studio debugger and see the allocator source. :)  

I'm using windows 7.  It seems to be trying to mmap my entire currently allocated virtual memory (16GB at the time, system managed), which just fails.  However, if I do some shenanigans (adding another 16gb paging file on a different drive, then removing it again, to result in 32gb allocated) to expand the allocated virtual memory in windows, while in a breakpoint after Heap_getMemoryLimit(), then allocating the large heap seems to succeed.    It seems that we can't ask windows to mmap the heap memory limit?

**Comment by muxanick** [comment]
It works for me in windows 10, I can ask for 16Gb twice, so seems Windows 7 has different behavior with that.

**Comment by Keshi** [comment]
The call to CreateFileMappingW seems to commit the entire amount right away.  I guess the difference is that I have 12GB of commit already from other programs.  The first call brings it to 26/32 Commit GB, and the second call fails.

I think you're right about the different behaviour.  According to https://blogs.msdn.microsoft.com/oldnewthing/20150130-00/?p=44793, it is "significantly improved in Windows 8.1, which defers committing the page tables until you actually use them."

**Comment by muxanick** [comment]
SEC_RESERVE is default I think on windows 10, but we can try to explicitly use it.
But you are right, GC shouln't ask for all available memory, it should have start up settings for that value

**Comment by Keshi** [comment]
According to https://blogs.technet.microsoft.com/markrussinovich/2008/11/17/pushing-the-limits-of-windows-virtual-memory/, it seems in at least one version of Windows, the behaviour of Window's commit limit expansion (when in system managed mode) is to expand it over time when the commit limit is approached.  I guess this explains why requesting so large a commit fails, as it doesn't approach the limit (just zooms past it in one go), and Windows will only succeed in raising the limit some time after the failure so it just rejects first.

**Comment by muxanick** [comment]
I submitted a change I think should fix win7.
I limited memory allocation to 2Gb (1Gb + 1Gb), it should be more than enough for the time until we figure out final version.
SEC_RESERVE is more complicated than I thought, we should have some callbacks from Immix GC to load virtual pages in memory (VirtualAlloc), that requires injection to Immix code and more testing. I will not able to do it anytime soon.
